File: chapters/05_retry_and_recovery/v3_adaptive_retry.ts
// TypeScript mirror for Chapter 05: Retry and Recovery (v3)
// Run with: npx ts-node v3_adaptive_retry.ts

type ErrorCategory = "transient" | "fatal" | "unknown";

interface LogEntry {
  event: string;
  value: string;
}

function classifyError(error: string): ErrorCategory {
  const lowered = error.toLowerCase();
  if (lowered.includes("timeout") || lowered.includes("network") || lowered.includes("429")) {
    return "transient";
  }
  if (lowered.includes("permission") || lowered.includes("auth") || lowered.includes("401") || lowered.includes("403")) {
    return "fatal";
  }
  if (lowered.includes("5xx") || lowered.includes("500") || lowered.includes("502") || lowered.includes("503") || lowered.includes("504")) {
    return "transient";
  }
  if (lowered.includes("4xx") || lowered.includes("400") || lowered.includes("404")) {
    return "fatal";
  }
  return "unknown";
}

function runWithAdaptiveRetry(
  simulatedErrors: string[],
  maxRetries: number = 4
): [string, LogEntry[]] {
  const logs: LogEntry[] = [];

  let attempt = 0;
  while (attempt <= maxRetries) {
    logs.push({ event: "attempt", value: String(attempt) });

    if (attempt < simulatedErrors.length) {
      const error = simulatedErrors[attempt];
      const category = classifyError(error);
      logs.push({ event: "error", value: error });
      logs.push({ event: "category", value: category });

      if (category === "fatal") {
        logs.push({ event: "decision", value: "stop" });
        return ["failed", logs];
      }

      if (attempt >= maxRetries) {
        logs.push({ event: "decision", value: "exhausted" });
        return ["failed", logs];
      }

      const backoffMs = Math.min(1000 * Math.pow(2, attempt), 8000);
      logs.push({ event: "retry_delay_ms", value: String(backoffMs) });
      logs.push({ event: "decision", value: "retry" });

      attempt++;
      continue;
    }

    logs.push({ event: "decision", value: "success" });
    return ["success", logs];
  }

  logs.push({ event: "decision", value: "exhausted" });
  return ["failed", logs];
}

function main(): void {
  const [status, logs] = runWithAdaptiveRetry(["network timeout", "network timeout"], 3);
  console.log(status);
  console.log(logs);
}

main();

File: chapters/05_retry_and_recovery/v3_adaptive_retry.py
from __future__ import annotations


def _classify(error: str) -> str:
    lowered = error.lower()
    if "timeout" in lowered or "network" in lowered or "429" in lowered:
        return "transient"
    if "permission" in lowered or "auth" in lowered or "401" in lowered or "403" in lowered:
        return "fatal"
    if "5xx" in lowered or "500" in lowered or "502" in lowered or "503" in lowered or "504" in lowered:
        return "transient"
    if "4xx" in lowered or "400" in lowered or "404" in lowered:
        return "fatal"
    return "unknown"


def run_with_adaptive_retry(simulated_errors: list[str], max_retries: int = 4) -> tuple[str, list[dict[str, str]]]:
    logs: list[dict[str, str]] = []

    attempt = 0
    while attempt <= max_retries:
        logs.append({"event": "attempt", "value": str(attempt)})

        if attempt < len(simulated_errors):
            error = simulated_errors[attempt]
            category = _classify(error)
            logs.append({"event": "error", "value": error})
            logs.append({"event": "category", "value": category})

            if category == "fatal":
                logs.append({"event": "decision", "value": "stop"})
                return "failed", logs

            if attempt >= max_retries:
                logs.append({"event": "decision", "value": "exhausted"})
                return "failed", logs

            backoff_ms = min(1000 * (2**attempt), 8000)
            logs.append({"event": "retry_delay_ms", "value": str(backoff_ms)})
            logs.append({"event": "decision", "value": "retry"})

            attempt += 1
            continue

        logs.append({"event": "decision", "value": "success"})
        return "success", logs

    logs.append({"event": "decision", "value": "exhausted"})
    return "failed", logs


def main() -> None:
    status, logs = run_with_adaptive_retry(["network timeout", "network timeout"], max_retries=3)
    print(status)
    print(logs)


if __name__ == "__main__":
    main()
